docs/csharp/programming-guide/concepts/expression-trees/how-to-use-expression-trees-to-build-dynamic-queries.md

```diff
@@ -1,107 +1,266 @@
+* pass the newly wrapped expression tree back into the provider for optional additional processing.
+
+For example, the source code for <xref:System.Linq.Queryable.Take> looks something like this:
+
+```csharp
+public static IQueryable<TSource> Take<TSource>(this IQueryable<TSource> source, int count) {
+
+    // wrap the original expression in a MethodCallExpression using the Call factory method    
+    var wrapped =
+        Expression.Call(
+            null,
+            CachedReflectionInfo.Take_TSource_2(typeof(TSource)),
+
+            // the original expression
+            source.Expression, 
+
+            Expression.Constant(count)
+        )
+
+    // pass the wrapped expression back to the provider
+    var newQuery = source.Provider.CreateQuery<TSource>(wrapped);
+
+    return newQuery;
+}
+```
+
+Thus, based on runtime state, you can conditionally replace the original query object with the result of these method calls, to get a new query.
+
+```csharp
+bool sortByLength = /* ... */;
+
+var qry = companyNamesSource;
+if (sortByLength) {
+    qry = qry.OrderBy(x => x.Length);
+}
+```
+
+## Varying the expression tree passed into the LINQ methods
+
+You can also choose to pass in different expressions to the LINQ methods, depending on runtime state:
+
+```csharp
+string startsWith = /* ... */;
+string endsWith = /* ... */;
+
+Expression<Func<string, bool>>? expr = null;
+if (!string.IsNullOrEmpty(startsWith) && !string.IsNullOrEmpty(endsWith)) {
+    expr = x => x.StartsWith(startsWith) || x.EndsWith(endsWith);
+} else if (!string.IsNullOrEmpty(startsWith)) {
+    expr = x => x.StartsWith(startsWith);
+} else if (!string.IsNullOrEmpty(endsWith)) {
+    expr = x => x.EndsWith(endsWith);
+} else {
+    expr = x => true;
+}
+
+var qry = companyNamesSource.Where(expr);
+```
+
+You might also want to compose the various sub-expressions using a third-party library such as [LinqKit](http://www.albahari.com/nutshell/linqkit.aspx)'s [PredicateBuilder](http://www.albahari.com/nutshell/predicatebuilder.aspx):
+
+```csharp
+// This is functionally equivalent to the previous example.
+
+string startsWith = /* ... */;
+string endsWith = /* ... */;
+
+Expression<Func<string, bool>>? expr = PredicateBuilder.New<string>(true);
+if (!string.IsNullOrEmpty(startsWith)) {
+    expr = expr.Or(x => x.StartsWith(startsWith));
+}
+if (!string.IsNullOrEmpty(endsWith)) {
+    expr = expr.Or(x => x.EndsWith(endsWith));
+}
+
+var qry = companyNamesSource.Where(expr);
+```
+
+## Constructing expression trees and queries using factory methods
+
+In all the examples up to this point, we've known the element type at compile time&mdash;`string`&mdash;and thus the type of the query&mdash;`IQueryable<string>`. But what if you want to add components to a query without limiting yourself to a specific element type? And moreover, what if the components you want to add are different depending on the type?
```

Comment:
restate as statements, not questions:

```suggestion
In all the examples up to this point, we've known the element type at compile time&mdash;`string`&mdash;and thus the type of the query&mdash;`IQueryable<string>`. You may need to add components to a query without limiting yourself to a specific element type. You may want to add are different components depending on the type.
```

Reply:
Fixed in cc6550ea6ce39020e8838d4ff08fb2fb59232a73